pkg/loop/internal/chain_reader.go

```diff
@@ -0,0 +1,286 @@
+package internal
+
+import (
+	"context"
+	"encoding/json"
```

Comment:
Has the new json package been tried? Do we intend to follow up on that? It seems like a very easy performance win, and would also allow us to calibrate how strict the decoding rules are.